### ISSUE TITLE
With using tabs in plugin settings, in Craft 3 they aren't formatted on main page

### ISSUE THREAD

**Issue by narration-sd** [craftcms/cms#2326]
### Description
For a complex plugin in Craft 2, I used tabs in its Settings dialog, via the cp's _include/forms.

Porting to Craft 3, the settings.html tabs don't have their tab labels formatted, or outlined and shadowed. Each tab just shows the text, though clicking that works. On the other pages used, the tabs look fine, and work fine. 

I patched this for the most part by manually including some missing CSS in the settings.html file. It's not complete, and not a stable solution if Craft's CP styling should change; thus this issue is raised. 

What's desired is that using tabs via _includes/forms should 'just work' on the settings.html main page, as it does on other pages producing all but the main tab (but see next about them extending _layouts/cp ). There are several -- their additional tabs show as you get deeper into the interface, avoiding first maintenance responders having to view and choose from complex diagnostic/integration pages.

The missing CSS would come from _layouts/cp, which the other  tabbed pages can extend from,, but this settings.html cannot - the page formatting will go entirely awry. I'm not  sure if it was moved from or was unnecessary for the tab formatting in Craft 2.

There's an underlying issue also to do with the special treatment of the settings page, where the text 'settings-' is appended to many items to make them available for form data retrieval, and other purposes. This is why the embedded CSS patch adds the settings- prefix to #tabs in its selectors, and so basic tab formatting works. Shadows are as yet missing, as noted.

Here are relevant parts of the settings.html code:

```{% set title %}
	Standards Agent - Settings
{% endset %}

{% import "_includes/forms" as forms %}
{% do view.registerAssetBundle("narrationsd\\standardsagent\\assetbundles\\SettingsAssetBundle") %}

<style>
  #main-container #main #main-content #content-container #settings-tabs {

    position: relative;
    z-index: 1;
    -webkit-box-shadow: inset 0 -1px 0 #e3e5e8;
    box-shadow: inset 0 -1px 0 #e3e5e8;
    min-height: 40px;
    overflow: hidden;

  }
  #main-container #main #main-content #content-container #settings-tabs ul {

    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    width: calc(100% + 1px);

  }
  ol, ul {

    list-style: none;

  }
  #main-container #main #main-content #content-container #settings-tabs ul li a.sel {

    color: #29323d;
    background: #fff;
    padding-bottom: 10px;
    cursor: default;

  }
  #main-container #main #main-content #content-container #settings-tabs ul li a {

    position: relative;
    display: block;
    padding: 10px 24px;
    padding-bottom: 10px;
    white-space: nowrap;
    overflow: hidden;

  }
  ::placeholder {

    color: #b9bfc6;

  }
  body.ltr #main-container #main #main-content #content-container #settings-tabs ul li a::after {

    right: 0;

  }
  body.ltr #main-container #main #main-content #content-container #settings-tabs ul li a::after {

    background: linear-gradient(to right, rgba(235, 237, 239, 0), #ebedef 17px);

  }

/* experiment unresolved: this next isn't enough to get the shadows back, 
    to complete tab formatting's  look on the settings.html page.  */

  body.ltr #main-container #main #main-content #content-container #settings-tabs ul li a.sel::after {

    background: linear-gradient(to right, rgba(255, 255, 255, 0), #fff 17px);

  }

</style>

{% set tabs = {
settings:            { label: "Settings"|t, url: url('standardsagent/settings') },
maintenance:         { label: "Maintenance"|t, url: url('standardsagent/maintenance') },
} %}

{% set selectedTab = 'settings' %}

...etc. page content...

```

### Additional info

PHP version	7.1.3-3+deb.sury.org~xenial+1
Database driver & version	MySQL 5.7.17-0ubuntu0.16.04.1
Image driver & version	Imagick 3.4.3RC2, ImageMagick 6.8.9-9
Craft edition & version	Craft Pro 3.0.0-RC4
Yii version	2.0.13.1
Twig version	2.4.4
Guzzle version	6.2.1
Imagine version	0.7-dev
Plugins
Bactrian	1.0.0
Trimmer	1.0.0
Video Embed	3.0.0
Element API	2.5.2
ImageOptimize	1.4.2
Standards Agent	2.0.1
Redactor	1.0.0.1

**Comment by narration-sd** [comment]
nb improved description as it's been a while since worked this; hope I have it precise enough to the case.

**Comment by narration-sd** [comment]
And here are some pictures:

First, tabs as  they  appear,  unformatted, on settings page without the patch

![tabs-unpatched](https://user-images.githubusercontent.com/247945/35175768-22dd71c8-fd2a-11e7-879d-e961c2bc7fbb.png)

Then, here is the appearance with the present incomplete patch (shaped and positioned, but  no shadows)

![tabs-present-patch](https://user-images.githubusercontent.com/247945/35175795-46ebab20-fd2a-11e7-9c55-7e89c9b422d6.png)

Finally, here is how it should look, and does look, from a non-settings.html page:

![tabs-other-page](https://user-images.githubusercontent.com/247945/35175853-8613c814-fd2a-11e7-932a-11dbc8cf7b30.png)

**Comment by brandonkelly** [comment]
I’m not against plugin settings having tabs, but in Craft 3 tabs should always be defined as a `tabs` variable on the template, and tab HTML rendering should be left up to the _layouts/cp.html template.

Unfortunately that `tabs` variable needs to be defined higher up in the template than plugins have access to via `settnigsHtml()`, which means you’d have to duplicate the `craft\base\Plugin::getSettingsResponse()` in your own plugin’s class, and add a new `tabs` variable in the `renderTemplate()` call.

I also just made it possible for plugins to modify variables that get passed to the template, via an event on `craft\web\View`, so as of RC7 you could go that route instead:

```php
use craft\web\View;
use craft\events\TemplateEvent;

public function init()
{
    parent::init();

    Event::on(View::class, View::EVENT_BEFORE_RENDER_TEMPLATE, function(TemplateEvent $e) {
        if (
            $e->template === 'settings/plugins/_settings' &&
            $e->variables['plugin'] === $this
        ) {
            // Add the tabs
            $e->variables['tabs'] = [
                ['label' => 'Settings', 'url' => '#tab-settings'],
                ['label' => 'Maintenance', 'url' => '#tab-maintenance'],
            ];
        }
    });
}
```

**Comment by narration-sd** [comment]
Ok, and thanks – appreciate your thinking, esp. that you got that event in there to try it, and I will look into this approach a  bit later, as the conversion isn’t a front of the queue project at moment.

Have you an easily arrived-at thought about getting the remaining shadow portion of the css patch to work for the present – vs. the commented attempt I made at the bottom of the css above??

C.

**Comment by brandonkelly** [comment]
If you go this route, the tabs will look like any other tabs in the Control Panel:

<img width="285" alt="screen shot 2018-01-19 at 3 51 32 pm" src="https://user-images.githubusercontent.com/47792/35176899-c1b32684-fd30-11e7-9901-c633bd1a1ffe.png">

From there you would just need to use this scaffolding in the template rendered by your `settingsHtml()` method:

```html
<div id="tab-settings">
    Settings tab contents
</div>
<div id="tab-maintenance" class="hidden">
    Maintenance tab contents
</div>
```

**Comment by narration-sd** [comment]
Ok, thanks again, and actually I prefer that appearance 😊

I should figure out the shadows thing anyway, so as not to fall farther behind on adequate css chops….

Enjoy  your weekend, Brandon – definitely earned  it!

I’m talking to the Composer dev in Netherlands who did the part with the winsymlinkjunctions – it’s very nasty in that pit…but have made a suggestion how to just avoid it, as he’s really off the project especially.

C.

**Comment by narration-sd** [comment]
Brandon, went ahead anyway, couldn't resist, and have this operating, all screens integrated, and 98% of the relevant javascript. Dinner beckons, however, and I had spent an interval decoding what you really wanted. There is always  something, for sure.

I had to comment out the check for `$e->variables['plugin']`, as this doesn't exist for the moment, and I didn't spot a likely candidate to replace it from what $e xdebugged has., 

Pic of the result,operating below, which I think has very nice contemporary and utile tabs -- maybe that other shadowed look came from leftovers??

There is a slight tweak in it for you guys also, as the hud shows a just-launched deleting records activity counting down, graphically and via the owned-job per-step status, which a certain pull request provides, when it does.  

Very happy with these results, sure gratitude your way, as one more thing falls off the list...!

![sa-counting-down](https://user-images.githubusercontent.com/247945/35179803-61bb0cd0-fd56-11e7-9c21-1ee8e853c6f6.png)
